cdhweb/pages/migration_utils.py

```diff
@@ -0,0 +1,154 @@
+import json
+import logging
+from datetime import datetime
+
+from django.core.serializers import serialize
+from django.db import connection
+from django.db.migrations.executor import MigrationExecutor
+from django.test import TestCase
+
+logger = logging.getLogger('wagtail.core')
+
+# see for an explanation of django-treebeard & the `path` attribute:
+# http://www.agilosoftware.com/blog/django-treebard-and-wagtail-page-creation/
+def get_parent(apps, page):
+    '''Find the parent of a wagtail page instance. Always returns the underlying
+    :class:`wagtail.core.models.Page` rather than a subclass.
+    '''
+
+    # if the path is 4 digits, we're at the root, so there is no parent
+    if len(page.path) == 4:
+        return None
+
+    # otherwise remove the final 4 digits to get the parent's id (pk)
+    Page = apps.get_model('wagtailcore', 'Page')
+    return Page.objects.get(path=page.path[:-4])
+
+# add_child utility method adapted from:
+# http://www.agilosoftware.com/blog/django-treebard-and-wagtail-page-creation/
+# see also:
+# https://github.com/wagtail/wagtail/blob/stable/2.3.x/wagtail/core/models.py#L442
+def add_child(apps, parent_page, klass, **kwargs):
+    '''Create a new draft wagtail page of type klass as a child of page instance
+    parent_page, passing along kwargs to its create() function.
+    '''
+
+    # add/set the correct content type for the new page
+    ContentType = apps.get_model('contenttypes.ContentType')
+    page_content_type = ContentType.objects.get_or_create(
+        model=klass.__name__.lower(),
+        app_label=klass._meta.app_label,
+    )[0]
+
+    # create the new page
+    page = klass.objects.create(
+        content_type=page_content_type,
+        path='%s%04d' % (parent_page.path, parent_page.numchild + 1),
+        depth=parent_page.depth + 1,
+        numchild=0,
+        live=False,  # create as a draft so that URL is set correctly on publish
+        **kwargs
+    )
+
+    # update the parent's child count
+    parent_page.numchild += 1
+    parent_page.save()
+
+    # log the creation and return the page
+    logger.info(
+        "Page created: \"%s\" id=%d content_type=%s.%s path=%s",
+        page.title,
+        page.id,
+        klass._meta.app_label,
+        klass.__name__,
+        page.url_path
+    )
+
+    return page
+
+# adapted from wagtail's `save_revision`:
+# https://github.com/wagtail/wagtail/blob/stable/2.3.x/wagtail/core/models.py#L631
+def create_revision(apps, page, user=None, created_at=datetime.now(), **kwargs):
```

Comment:
Forgot to ask — what do we need revisions for?  Are they required for content to display?

Reply:
revisions associate text content with the page model; it's a one -> many. pages themselves have no content stored in the database and you need at least one revision in order for a page to have content.